File: src/core/confidence.ts
import type { Classification, ConfidenceLevel } from '../types.js';

// Sigmoid calibration — tuned so that:
// - Raw score 0.35+ maps to "ai-generated" territory
// - Raw score 0.20-0.35 maps to "unknown"
// - Raw score < 0.20 maps to "human-written"
// This is intentionally generous because individual detectors
// already apply conservative thresholds internally.
const SIGMOID_K = 8;
const SIGMOID_MIDPOINT = 0.40;

// Classification thresholds on RAW score (before sigmoid)
const AI_THRESHOLD = 0.42;      // > 0.42 = ai-generated
const HUMAN_THRESHOLD = 0.25;   // < 0.25 = human-written

/**
 * Map a raw 0-1 score through sigmoid to calibrated 0-100 confidence.
 */
export function calibrateConfidence(rawScore: number): number {
  const sigmoid = 1 / (1 + Math.exp(-SIGMOID_K * (rawScore - SIGMOID_MIDPOINT)));
  return Math.round(sigmoid * 100);
}

/**
 * Classify based on raw score thresholds.
 */
export function classifyFromScore(rawScore: number): Classification {
  if (rawScore >= AI_THRESHOLD) return 'ai-generated';
  if (rawScore <= HUMAN_THRESHOLD) return 'human-written';
  return 'unknown';
}

/**
 * Determine overall confidence level from line ranges.
 */
export function overallConfidenceLevel(
  ranges: readonly { confidence: number; classification: string; startLine: number; endLine: number }[],
  totalLines: number,
): ConfidenceLevel {
  if (ranges.length === 0) return 'LOW';

  let totalConfidence = 0;
  let unknownLines = 0;

  for (const range of ranges) {
    const rangeLines = range.endLine - range.startLine + 1;
    totalConfidence += range.confidence * rangeLines;
    if (range.classification === 'unknown') {
      unknownLines += rangeLines;
    }
  }

  const avgConfidence = totalConfidence / totalLines;
  const unknownPct = unknownLines / totalLines;

  if (avgConfidence > 65 && unknownPct <= 0.20) return 'HIGH';
  if (avgConfidence >= 40 && unknownPct <= 0.40) return 'MEDIUM';
  return 'LOW';
}

File: code_provenance/core/confidence.py
"""Confidence calibration and classification from raw scores."""

from __future__ import annotations

import math

from code_provenance.types import Classification, ConfidenceLevel

# Sigmoid calibration -- tuned so that:
# - Raw score 0.35+ maps to "ai-generated" territory
# - Raw score 0.20-0.35 maps to "unknown"
# - Raw score < 0.20 maps to "human-written"
SIGMOID_K = 8
SIGMOID_MIDPOINT = 0.40

# Classification thresholds on RAW score (before sigmoid)
AI_THRESHOLD = 0.42  # > 0.42 = ai-generated
HUMAN_THRESHOLD = 0.25  # < 0.25 = human-written


def calibrate_confidence(raw_score: float) -> int:
    """Map a raw 0-1 score through sigmoid to calibrated 0-100 confidence."""
    sigmoid = 1 / (1 + math.exp(-SIGMOID_K * (raw_score - SIGMOID_MIDPOINT)))
    return round(sigmoid * 100)


def classify_from_score(raw_score: float) -> Classification:
    """Classify based on raw score thresholds."""
    if raw_score >= AI_THRESHOLD:
        return Classification.AI_GENERATED
    if raw_score <= HUMAN_THRESHOLD:
        return Classification.HUMAN_WRITTEN
    return Classification.UNKNOWN


def overall_confidence_level(
    ranges: list[dict],
    total_lines: int,
) -> ConfidenceLevel:
    """Determine overall confidence level from line ranges.

    Args:
        ranges: List of dicts with 'confidence', 'classification',
                'start_line', 'end_line' keys.
        total_lines: Total number of lines in the file.
    """
    if not ranges:
        return ConfidenceLevel.LOW

    total_confidence = 0.0
    unknown_lines = 0

    for r in ranges:
        range_lines = r["end_line"] - r["start_line"] + 1
        total_confidence += r["confidence"] * range_lines
        if r["classification"] == "unknown":
            unknown_lines += range_lines

    avg_confidence = total_confidence / total_lines
    unknown_pct = unknown_lines / total_lines

    if avg_confidence > 65 and unknown_pct <= 0.20:
        return ConfidenceLevel.HIGH
    if avg_confidence >= 40 and unknown_pct <= 0.40:
        return ConfidenceLevel.MEDIUM
    return ConfidenceLevel.LOW
